Migrate index_add cpu from TH to ATen

ggoossen opened the pull request:
Summary:
Migrate index_add cpu from TH to ATen.

I couldn't find replacement for get1d and set1d, so doing pointer arithmetic inplace.

Test Plan: existing tests

Differential Revision: D18060971

gchanan commented:
CC @VitalyFedyunin

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+    }
+  }
+  else {
+    TORCH_CHECK_INDEX(source.dim() <= 1, "source tensor must have one dimension");
```
nit: I'd just fix these error messages, i.e. here say "source tensor must have 0 or 1 dimensions".

ggoossen replied:
I've changed the error messages, although I'm not sure this is the right direction.
Documentation for `index_add_` states `index` must be a vector., and talks about the `dim`th dimension of `tensor`, so this support of scalar `index` and `tensor` contradicts the documentation. Also the operator isn't very useful when `tensor` / `index` is a scalar. I suspect the support for scalars is mostly an artifact of TH api.

I would prefer deprecating scalars for `index` and `source` it instead of "documenting" it in the error messages.

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+    TORCH_CHECK_INDEX(source.dim() <= 1, "source tensor must have one dimension");
+    TORCH_CHECK_INDEX(
+        numel == source.numel(),
+        "index_add_(): Number of indices (", numel, ") should be equal to source.size(dim) (", source.numel(), ")");
```
here too, just say "should be equal to source.numel".

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+    AT_DISPATCH_ALL_TYPES(self.scalar_type(), "index_add_", [&] {
+      auto self_stride = self.dim() == 0 ? 1 : self.stride(dim);
+      auto source_stride = source.dim() == 0 ? 1 : source.stride(dim);
+      for (auto i=0; i<numel; i++) {
```
nit: be consistent about spacing around operators, just add a couple of spaces here.

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+  auto index_data = index_contig.data_ptr<int64_t>();
+
+  if (self.dim() > 1) {
+    for (auto i=0; i<numel; i++) {
```
nit: be consistent about spacing around operators, just add a couple of spaces here.

gchanan reviewed (commented):
looks pretty good, just some nits.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3755,7 +3755,7 @@
   variants: method
   dispatch:
-    CPU: legacy::cpu::_th_index_add_
+    CPU: index_add__cpu
```
nit: I think it's clearer and more consistent to name this "index_add_cpu_", see e.g. "bernoulli_scalar_cpu_".  It's difficult to discern "__" and more consistent to always just end in an underscore.

ggoossen replied:
__cpu is also used in a lot of placed. But I agree the .._cpu_ is better. Maybe worth codemodding the others?

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+  }
+  else {
+    TORCH_CHECK_INDEX(source.dim() <= 1, "source tensor must have one or zero dimensions");
+    TORCH_CHECK_INDEX(
```
why is this not above the if/else?

ggoossen replied:
source.dim should be equal to self.dim (with 0 and 1 being equal), but the original code didn't check that either, only in this else branch the source.dim was check to be 0/1.

Ideally we should check source.dim and self.dim are equal, but the original code didn't do that, so I thought changing that in the migration diff wouldn't be the best idea. Or do you think we should add that in this diff anyway? Which would also mean we could give a much cleaner error message.

gchanan replied:
Good question.  My concern is we aren't checking:
```THArgCheck(numel == THTensor_sizeLegacyNoScalars(src, dim),4,"Number of indices should be equal to source:size(dim)");
  THArgCheck(dim < THTensor_nDimensionLegacyNoScalars(src), 4,"Indexing dim %d is out of bounds of tensor", dim);	
  THArgCheck(numel == THTensor_sizeLegacyNoScalars(src, dim),4,"Number of indices should be equal to source:size(dim)");	

```

in the if.

I think just leaving those checks as is seems reasonable (which is above the if) -- what do you think?

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+    }
+  }
+  else {
+    TORCH_CHECK_INDEX(source.dim() <= 1, "source tensor must have one or zero dimensions");
```
why is this TORCH_CHECK_INDEX?  This is checking the source?

gchanan replied:
Also this error message looks confusing, because ""source tensor must have one or zero dimensions"" is only true in the else, right?

ggoossen replied:
changed to TORCH_CHECK and changed error message (but ideally we'll change to check source.dim == self.dim).

albanD commented:
Do you have any benchmark to compare the runtime of these two implementations?

ggoossen commented:
@albanD I didn't run any benchmarks. Are there existing benchmarks to run?

albanD commented:
I'm afraid no :/
But a simple `%timeit` for few common input sizes should give an ok signal.

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+
+  auto numel = index.numel();
+  TORCH_CHECK_INDEX(index.dim() <= 1, "index_add_(): Index is supposed to be a vector");
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "index_add_(): Expected LongTensor for index");
```
nit: better to say "expected dtype int64 for index".

gchanan commented on aten/src/ATen/native/Indexing.cpp:
```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
+    for (auto i = 0; i < numel; i++) {
+      auto selfSlice = self.select(dim, index_data[i]);
+      auto sourceSlice = source.select(dim, i);
+      selfSlice.add_(sourceSlice, 1);
```
nit: do you need the 1?

ggoossen commented:
Did some timing:

On 1d tensors:
before: 3.72
after: 1.57

On 2d tensors:
before: 0.57
after: 3.16

So it is more than 2x faster for 1d tensors (probably due to some check from get1d and set1d now only being done once).
But for 2d tensors it is about 5x slower, probably the .select is quite a bit slower than (select_), it is probably worth optimizing this. But probably better to do that after this migration.

Timed using:
```
import timeit
print(timeit.timeit(
    "dst.index_add_(index=index, dim=0, source=value)",
    setup="""                                                                                                                                                                                                                                                                             
import torch                                                                                                                                                                                                                                                                              
dst = torch.rand((1024, 32))                                                                                                                                                                                                                                                              
index = torch.randint(1024, (16*1024, ))                                                                                                                                                                                                                                                  
value = torch.rand((16*1024, 32))                                                                                                                                                                                                                                                         
    """,
    number=10,
))

print(timeit.timeit(
    "dst.index_add_(index=index, dim=0, source=value)",
    setup="""                                                                                                                                                                                                                                                                             
import torch                                                                                                                                                                                                                                                                              
dst = torch.rand((1024*32))                                                                                                                                                                                                                                                               
index = torch.randint(1024*32, (16*1024*32, ))                                                                                                                                                                                                                                            
value = torch.rand((16*1024*32))                                                                                                                                                                                                                                                          
    """,
    number=10
))
```

ggoossen commented:
Failing test is:

Nov 12 18:42:03 FAIL: test_embedding_bag_device_cuda_float32 (__main__.TestNNDeviceTypeCUDA)
...
1890 | Nov 12 18:42:03 AssertionError: tensor(2.00272e-05, device='cuda:0', dtype=torch.float32) not less than or equal to 2e-05 :

Which is the cuda version, which isn't modified here. Allowed error margin should be increased for the test.

gchanan commented:
code looks correct.

The performance difference looks like it's caused by the lack of `select_out` -- IMO we should verify that but adding select_out and seeing if it erases the perf difference.

ggoossen commented:
@gchanan I looked at adding `select_out`, but that looked a bit complicated in that it required _out version of a lot of other functions. I also did some additional profiling and also creating the TensorIterator in add_ seemed expensive.
So changed index_add_ to create the TensorIterator once, and only update the operand data pointer as the strides and dims remain.
This seems to work quite fast, and new code is about 3-5 times faster then the existing code.
time on 2d tensor:
before: 1.6310803790111095
after: 0.3265240238979459
time on 1d tensor:
11.638618602184579
3.293580493191257

ggoossen commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -231,8 +231,8 @@ struct CAFFE2_API TensorIterator {
-  /// Replaces the data pointer and strides for the operand at index `arg`
-  void replace_operand(int arg, void* data, IntArrayRef stride);
+  /// Replaces the data pointer for the operand at index `arg`
+  void replace_operand(int arg, void* data);
```
This function wasn't used before. Changing the `stride` seems wrong, as the stride needs to be in sync with the dimensions, and the original values could have been reordered/coalesced in build.

VitalyFedyunin reviewed (commented):
Looks correct with couple nits:
1. please rename `replace_operand` to `unsafe_replace_operand`
2. we have zero test coverage of `index_add_` with non-contiguous tensors, taking inso account that you are using pointer math, it would be nice to add such tests.

VitalyFedyunin reviewed (commented):
Looks correct with couple nits:
1. please rename `replace_operand` to `unsafe_replace_operand`
2. we have zero test coverage of `index_add_` with non-contiguous tensors, taking inso account that you are using pointer math, it would be nice to add such tests.

ggoossen commented:
> 2. we have zero test coverage of `index_add_` with non-contiguous tensors, taking inso account that you are using pointer math, it would be nice to add such tests.

@VitalyFedyunin These would indeed nice to have. Is there a standard way of adding these? I would expect some framework where we provided a reference implementation, and some arg specification, based on which these tests are generated. `common_nn` seems to have this for modules, is there something similar for functions?

VitalyFedyunin commented:
I have no good reference to copy `index_add_` test from top of my head, but looking into `test/test_torch.py` will likely help you to find something similar.

ggoossen committed:
Migrate index_add cpu from TH to ATen (#28421)

Summary:
Migrate index_add cpu from TH to ATen.

I couldn't find replacement for get1d and set1d, so doing pointer arithmetic inplace.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/28421

Test Plan: existing tests

Differential Revision: D18060971

Pulled By: ggoossen

fbshipit-source-id: 85fba294efc8bf00c64f7e79fe1081cbc15f64a4

gchanan commented:
just a nit on the naming: `unsafe_replace_operand` it's better to put unsafe at the end, often people will want to see the safe/unsafe versions of functions, not "show me all the unsafe functions".

VitalyFedyunin commented:
Good point, I plan to rewrite `stack` operator and will rename this function in the process.
